### PR TITLE
[Kernel] Add basic support for `checkpointProtection` by limiting log cleanup

### PR DESCRIPTION
**Pull request by dengsh12**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description
Add basic support for `checkpointProtection` table feature in Kernel by limiting log cleanup after checkpointing. Only do log cleanup when
1. Snapshot was built as "latest" by intent (not time-traveled)
2. checkpointProtection feature is not enabled
3. delta.enableExpiredLogCleanup table property is set to true

## How was this patch tested?
Added tests

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 1]
```diff
@@ -390,6 +390,25 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+  test("try create a checkpoint when checkpointProtection enabled") {
+    withTempDirAndEngine { (tablePath, tc) =>
+      // Create a table with checkpointProtection feature enabled
+      spark.sql(s"CREATE TABLE delta.`$tablePath` (id INT) USING delta " +
+        "TBLPROPERTIES ('delta.feature.checkpointProtection' = 'supported')")
+
+      // Add some data to create commits
+      for (_ <- 0 to 3) {
+        spark.sql(s"INSERT INTO delta.`$tablePath` VALUES (1), (2), (3)")
+      }
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: do you need to use delta-spark? Couldn't you use kernel to create the table? and look at other tests for how we append/insert data?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 2]
```diff
@@ -390,6 +390,25 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+
+      // Attempting to checkpoint should fail
+      val ex = intercept[Exception] {
+        kernelCheckpoint(tc, tablePath, checkpointVersion = 4)
+      }
+      assert(ex.getMessage.contains("checkpointProtection"))
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Assert more of the error message please. this makes reading this single test even cleaner.

**Comment by dengsh12** [comment]
> Awesome! Nice and short and simple. Thank you! A few comments on the test -- After those are addressed, will stamp 🚀

Thanks for the suggestions, really good points! Fixed the test, still pending the discussion about a corner case

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java [hunk 3]
```diff
@@ -106,9 +107,19 @@ public static void checkpoint(Engine engine, Clock clock, SnapshotImpl snapshot)
     logger.info(
         "{}: Finished writing last checkpoint metadata file for version: {}", tablePath, version);
 
-    // Clean up delta log files if enabled.
+    // Only clean up expired log files when:
+    // 1. Snapshot was built as "latest" by intent (not time-traveled)
+    // 2. checkpointProtection feature is not enabled
+    // 3. delta.enableExpiredLogCleanup table property is set to true
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java]: Could this just become a helper function called `shouldPerformLogCleanup`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 4]
```diff
@@ -142,6 +150,15 @@ public long getVersion() {
     return version;
   }
 
+  /**
+   * Returns true if this snapshot was built as a "latest" snapshot query (i.e., no time-travel
+   * parameters were provided). This is intent-based - it indicates what the user requested, not
+   * whether the snapshot is actually the latest version.
+   */
+  public boolean wasBuiltAsLatest() {
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java]: This is in the `// Public APIs //` section, yet this is not a public Snapshot API. This is a public internal API on the SnapshotImpl. Please move down to the right section?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 5]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+
+      // Create table (version 0) and add commits (version 1, 2, 3)
+      appendData(
+        engine,
+        tablePath,
+        isNewTable = true,
+        schema = testSchema,
+        data = data,
+        tableProperties = tableProperties)
+      for (_ <- 1 to 3) { appendData(engine, tablePath, data = data) }
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: nit: add a newline before this. this would make this test flow a bit more readable

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 6]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+
+      val deltaLogDir = new File(tablePath, "_delta_log")
+      assert(deltaLogDir.listFiles().count(_.getName.endsWith(".json")) === 4)
+
+      // Checkpoint at version 2 using SnapshotBuilder.atVersion() - wasBuiltAsLatest=false
+      val clock = new KernelClock { def getTimeMillis: Long = System.currentTimeMillis() }
+      val snapshot = TableManager.loadSnapshot(tablePath)
+        .atVersion(2).build(engine).asInstanceOf[SnapshotImpl]
+      Checkpointer.checkpoint(engine, clock, snapshot)
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: why not `snapshot.checkpoint(..)` ?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 7]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
       verifyResults(tablePath, expResults, checkpointVersion)
     }
   }
+
+  test("log cleanup: non-latest snapshot does NOT trigger log cleanup") {
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: snapshots don't inherintely `do` or `not do` log cleanup
we are more so testing what they are `capable` of doing

this shold be `non-latest snapshot cannot trigger ..`

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 8]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+      Checkpointer.checkpoint(engine, clock, snapshot)
+
+      // Verify no log cleanup happened
+      assert(
+        deltaLogDir.listFiles().count(_.getName.endsWith(".json")) === 4,
+        "Checkpoint on snapshot built with specific version should NOT trigger log cleanup")
+    }
+  }
+
+  test("log cleanup: latest snapshot triggers log cleanup") {
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: `latest snapshot can trigger`

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 9]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+        engine,
+        tablePath,
+        isNewTable = true,
+        schema = testSchema,
+        data = data,
+        tableProperties = tableProperties)
+      for (_ <- 1 to 3) { appendData(engine, tablePath, data = data) }
+
+      val deltaLogDir = new File(tablePath, "_delta_log")
+      assert(deltaLogDir.listFiles().count(_.getName.endsWith(".json")) === 4)
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Is this code the same as above? Are we able to consolidate it? Maybe not -- but try/consider that

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 10]
```diff
@@ -390,6 +394,117 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+      Checkpointer.checkpoint(engine, clock, latestSnapshot)
+
+      // Verify log cleanup happened
+      assert(
+        deltaLogDir.listFiles().count(_.getName.endsWith(".json")) < 4,
+        "Checkpoint on snapshot built without specific version should trigger log cleanup")
+    }
+  }
+
+  test("log cleanup: checkpointProtection enabled prevents log cleanup") {
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: `.. prevents log cleanup, even at least snapshot version`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 11]
```diff
@@ -115,6 +122,7 @@ public SnapshotImpl(
     this.metadata = requireNonNull(metadata);
     this.committer = committer;
     this.inCommitTimestampOpt = inCommitTimestampOpt;
+    this.wasBuiltAsLatest = snapshotContext.isLatestQuery();
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:128]: Nice! I think this is broken for post-commit snapshots. Can you please add a todo?

That is, in a post-commit snapshot, we _are_ in fact the latest Snapshot (because we just wrote!) -- yet this will say that we are _not_ the latest snapshot _probably_.

I believe that post-commit-snapshot just isn't baked into the SnapshotQueryContext yet

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java (hunk 3 above)
Comment (dengsh12) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java]: Good point! Moved

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java (hunk 4 above)
Comment (dengsh12) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java]: Thanks for letting me know this! Didn't notice that at all before

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 5 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Done!

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 6 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Good point! Switched to use `snapshot.writeCheckpoint`

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 7 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Fixed

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 8 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Fixed

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 9 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Good point! Refactored.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 10 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Fixed. As the `latest` comes from user's intent, changed to `... prevents log cleanup, even snapshot is built as latest`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java [hunk 12]
```diff
@@ -106,16 +107,34 @@ public static void checkpoint(Engine engine, Clock clock, SnapshotImpl snapshot)
...
 
+  /**
+   * Only clean up expired log files when:
+   *
+   * 1. Snapshot was built as "latest" by intent (not time-traveled)
+   * 2. checkpointProtection feature is not enabled
+   * 3. delta.enableExpiredLogCleanup table property is set to true
+   */
+  private static boolean shouldPerformLogCleanup(SnapshotImpl snapshot) {
+    final Metadata metadata = snapshot.getMetadata();
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java]: nit: you could probably delete this single-use variable?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala [hunk 13]
```diff
@@ -390,6 +422,73 @@ class CreateCheckpointSuite extends CheckpointBase {
...
+  test("log cleanup: latest snapshot can trigger log cleanup") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val commits = 3
+      val tableProperties = Map(
+        "delta.logRetentionDuration" -> "interval 0 seconds",
+        "delta.enableExpiredLogCleanup" -> "true")
+      val deltaLogDir = setupTestTable(engine, tablePath, tableProperties, commits)
+
+      // Get latest snapshot (version == commits) using builder without atVersion() - wasBuiltAsLatest=true
+      val latestSnapshot = TableManager.loadSnapshot(tablePath).build(engine)
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: nit: scala style would probably prefer this written as

```scala
val latestSnapshot = TableManager.loadSnapshot(tablePath)
  .build(engine)
  .asInstanceOf[SnapshotImpl]
```

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java (hunk 11 above)
Comment (dengsh12) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:128]: Added todo

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java (hunk 12 above)
Comment (dengsh12) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java]: Fixed

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala (hunk 13 above)
Comment (dengsh12) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CreateCheckpointSuite.scala]: Gotta! Applied
